### PR TITLE
[sqlite] add compatible async-storage

### PR DESCRIPTION
**Pull request by Kudo**
# Why

close ENG-12428

# How

add key-value storage and offers compatible with react-native-async-storage

# Test Plan

- doc preview at /versions/unversioned/sdk/sqlite/#asyncstorage
- add unit tests and ci passed
- apply the patch to test NCL
```diff
--- a/apps/native-component-list/src/screens/AsyncStorageScreen.tsx
+++ b/apps/native-component-list/src/screens/AsyncStorageScreen.tsx
@@ -1,4 +1,4 @@
-import AsyncStorage from '@react-native-async-storage/async-storage';
+import AsyncStorage from 'expo-sqlite/async-storage';
 import * as Updates from 'expo-updates';
 import * as React from 'react';
 import { Button, StyleSheet, Text, View } from 'react-native';
```

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by Kudo** [comment]
* **#31596** <a href="https://app.graphite.dev/github/pr/expo/expo/31596?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Kudo and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/expo-sqlite/src/Storage.ts [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+/**
+ * This default instance of the [`SQLiteStorage`](#sqlitestorage-1) class is used as a drop-in replacement for the `AsyncStorage` module from [`@react-native-async-storage/async-storage`](https://github.com/react-native-async-storage/async-storage).
+ */
+export const AsyncStorage = new SQLiteStorage('ExpoSQLiteStorage');
+
+export default AsyncStorage;
+
+/**
+ * Alias for [`AsyncStorage`](#asyncstorage-1), given the storage not only offers asynchronous methods.
```
Comment (amandeepmittal) [inline packages/expo-sqlite/src/Storage.ts:390]:
```suggestion
 * Alias for [`AsyncStorage`](#sqliteasyncstorage), given the storage not only offers asynchronous methods.
```

Broken internal link.

Comment (Kudo) [reply packages/expo-sqlite/src/Storage.ts:390]: good catch thanks 💛

**Review by alanjhughes — APPROVED** [review]
This is great @Kudo! Just a few nit comments on wording

File: packages/expo-sqlite/src/Storage.ts [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+  private db: SQLiteDatabase | null = null;
+
+  constructor(private readonly databaseName: string) {}
+
+  //#region Asynchronous API
+
+  /**
+   * Retrieves the value associated with the given key asynchronously.
+   */
+  getItemAsync = async (key: string): Promise<string | null> => {
```
Comment (tsapeta) [inline packages/expo-sqlite/src/Storage.ts:34]: I would be more in favor of using function syntax over arrow functions. If you don't need to capture `this` from the outer scope, use a function.

Also, assigning an arrow function to a property like this is not adding the function into the prototype, but to each instance (slower object initialization, higher memory consumption).

Comment (Kudo) [reply packages/expo-sqlite/src/Storage.ts:34]: i was thinking if people calling the storage functions in callbacks and the `this` being overrode. maybe it's an unnecessary pre-optimization. anyhow, i've reverted the changes in https://github.com/expo/expo/pull/31676

File: docs/pages/versions/unversioned/sdk/sqlite.mdx [hunk 3]
```diff
@@ -409,3 +440,4 @@ The following table summarizes the common API for [`SQLiteDatabase`](#sqlitedata
 | [`getEachAsync()`](#geteachasyncsource-params)   | [`executeAsync()`](#executeasyncparams) + `for-await-of` async iterator       | Provides an iterator for result set traversal. This method fetches one row at a time from the database, potentially reducing memory usage compared to `getAllAsync()`. | Recommended for handling large result sets incrementally, such as with infinite scrolling implementations.                                                                                 |
 
 <APISection packageName="expo-sqlite" apiName="SQLite" exposeAllClassPropsInSidebar />
+```
```
[the remark this answers is not in the document]
Comment (Kudo) [reply docs/pages/versions/unversioned/sdk/sqlite.mdx:443]: good catch and thanks for this great post-review 💛

**Pull request merged**